From c7d3993b5f37954738e649eb9e52e28546c4a1b7 Mon Sep 17 00:00:00 2001
From: Icarus Sparry <icarus.w.sparry@intel.com>
Date: Tue, 22 Aug 2017 10:25:00 -0700
Subject: [PATCH 1/2] Gcc 7 fixes

gcc 7 adds some more diagnostics, and as we compile with diagnostics
becoming errors this causes us to fail to build. The fixes are
increasing a buffer size so gcc knows it will not overflow, adding in
multiple /* Falls through */ comments to document that the lack of
break statement is intentional, and adding gcc pragmas to indicate
that it is deliberate that we don't need the terminating NUL character
in a character array inside a struct.

Redhat have a different approach, they just turn off the warnings in
the Makefile! If their changes get accepted upstream then this patch
should probably be removed. However I think it is better to keep as
many warnings enabled as possible and just mark the false positives.

Signed-off-by: Icarus Sparry <icarus.w.sparry@intel.com>

Updated 2018-10-29:

Removed the pragmas and the buffer size fix, since they were addressed
in the 4.1 release.

Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>
---
 Grow.c        | 4 ++++
 bitmap.c      | 8 ++++++++
 mdadm.c       | 2 ++
 super-intel.c | 1 +
 util.c        | 1 +
 5 files changed, 16 insertions(+)

diff --git a/Grow.c b/Grow.c
index 4436a4d..052866a 100644
--- a/Grow.c
+++ b/Grow.c
@@ -1447,6 +1447,7 @@ char *analyse_change(char *devname, struct mdinfo *info, struct reshape *re)
 		switch (info->new_level) {
 		case 4:
 			delta_parity = 1;
+			/* Falls through */
 		case 0:
 			re->level = 4;
 			re->before.layout = 0;
@@ -1474,10 +1475,12 @@ char *analyse_change(char *devname, struct mdinfo *info, struct reshape *re)
 
 	case 4:
 		info->array.layout = ALGORITHM_PARITY_N;
+		/* Falls through */
 	case 5:
 		switch (info->new_level) {
 		case 0:
 			delta_parity = -1;
+			/* Falls through */
 		case 4:
 			re->level = info->array.level;
 			re->before.data_disks = info->array.raid_disks - 1;
@@ -1533,6 +1536,7 @@ char *analyse_change(char *devname, struct mdinfo *info, struct reshape *re)
 		case 4:
 		case 5:
 			delta_parity = -1;
+			/* Falls through */
 		case 6:
 			re->level = 6;
 			re->before.data_disks = info->array.raid_disks - 2;
diff --git a/bitmap.c b/bitmap.c
index e38cb96..179a019 100644
--- a/bitmap.c
+++ b/bitmap.c
@@ -82,13 +82,21 @@ static inline int count_dirty_bits_byte(char byte, int num_bits)
 
 	switch (num_bits) { /* fall through... */
 		case 8:	if (byte & 128) num++;
+			/* Falls through */
 		case 7:	if (byte &  64) num++;
+			/* Falls through */
 		case 6:	if (byte &  32) num++;
+			/* Falls through */
 		case 5:	if (byte &  16) num++;
+			/* Falls through */
 		case 4:	if (byte &   8) num++;
+			/* Falls through */
 		case 3: if (byte &   4) num++;
+			/* Falls through */
 		case 2:	if (byte &   2) num++;
+			/* Falls through */
 		case 1:	if (byte &   1) num++;
+			/* Falls through */
 		default: break;
 	}
 
diff --git a/mdadm.c b/mdadm.c
index 25a1abd..3f4eac5 100644
--- a/mdadm.c
+++ b/mdadm.c
@@ -155,6 +155,7 @@ int main(int argc, char *argv[])
 			    mode == CREATE || mode == GROW ||
 			    mode == INCREMENTAL || mode == MANAGE)
 				break; /* b means bitmap */
+			/* Falls through */
 		case Brief:
 			c.brief = 1;
 			continue;
@@ -839,6 +840,7 @@ int main(int argc, char *argv[])
 
 		case O(INCREMENTAL,NoDegraded):
 			pr_err("--no-degraded is deprecated in Incremental mode\n");
+			/* Falls through */
 		case O(ASSEMBLE,NoDegraded): /* --no-degraded */
 			c.runstop = -1; /* --stop isn't allowed for --assemble,
 					 * so we overload slightly */
diff --git a/super-intel.c b/super-intel.c
index 6438987..47434d3 100644
--- a/super-intel.c
+++ b/super-intel.c
@@ -3425,6 +3425,7 @@ static void getinfo_super_imsm_volume(struct supertype *st, struct mdinfo *info,
 
 			}
 		}
+			/* Falls through */
 		case MIGR_VERIFY:
 			/* we could emulate the checkpointing of
 			 * 'sync_action=check' migrations, but for now
diff --git a/util.c b/util.c
index c26cf5f..49ce717 100644
--- a/util.c
+++ b/util.c
@@ -400,6 +400,7 @@ unsigned long long parse_size(char *size)
 		switch (*c) {
 		case 'K':
 			c++;
+			/* Falls through */
 		default:
 			s *= 2;
 			break;
-- 
2.19.1

